### ISSUE TITLE
Does not support Apple Silicon

### ISSUE THREAD

**Issue by fxcoudert** [IoLanguage/io#465]
Compilation on ARM macOS (aka Apple Silicon) leads to a build failure:

```
/tmp/io-20221111-68531-j5uskf/io-2017.09.06/libs/coroutine/source/asm.S:143:2: error: unrecognized instruction mnemonic
 mflr r0
 ^
/tmp/io-20221111-68531-j5uskf/io-2017.09.06/libs/coroutine/source/asm.S:144:2: error: unrecognized instruction mnemonic
 mfcr r5
 ^
/tmp/io-20221111-68531-j5uskf/io-2017.09.06/libs/coroutine/source/asm.S:145:2: error: unrecognized instruction mnemonic
 mfctr r6
 ^
/tmp/io-20221111-68531-j5uskf/io-2017.09.06/libs/coroutine/source/asm.S:146:2: error: unrecognized instruction mnemonic
 mfxer r7
 ^
/tmp/io-20221111-68531-j5uskf/io-2017.09.06/libs/coroutine/source/asm.S:147:13: error: unexpected token in argument list
 stw r0, 0*4(r3)
            ^
```

**Comment by shayneoneill** [comment]
Any chance of getting this fixed? Intel macs are getting closer and closer to EOLed for support, and with that usually comes the end of Rosetta.

**Comment by PedroBatista** [comment]
I think Rosetta will be around for many years, the problem is ( seems to be ) the assembly code in asm.S. It also blocked me from running this with WASM.

Since assembly programming is a long lost art, my guess would be to tell ChatGPT to translate it ( unlikely ) or to implement #define NEEDARMCONTEXT 1
#define SET setmcontext
#define GET getmcontext 

in C.

**Comment by stevedekorte** [comment]
I thought Apple finally got ucontext working. If so, we can just set the ifdef for that.

**Comment by stevedekorte** [comment]
> I think Rosetta will be around for many years, the problem is ( seems to be ) the assembly code in asm.S. It also blocked me from running this with WASM.

There is an IoWASM project: https://github.com/stevedekorte/ioWasm

And an online REPL demo that uses it: https://iolanguage.org/repl/index.html

Would be great to see some bindings into JS and the DOM.

**Comment by shayneoneill** [comment]
@PedroBatista Perhaps, but the original Rosetta (for running PowerPC apps on Intel macs) really only was around for about 4-5 years, and we're at year 3 with Rosetta2. So maybe a couple of more years? Its a *bit* of breathing room, but not much.

That said, Apple will likely give some deprecation notice a version or two of macos before yeeting it.

**Comment by PedroBatista** [comment]
Just an heads-up, I've managed to make changes to the code and compile it natively to Apple Silicon. But I've been very busy, hopefully I'll do a PR soon.

**Comment by PedroBatista** [comment]
@stevedekorte Yes, I remember I run the existing tests ( iovm/tests/correctness/CoroTest.io - I think? It was a couple months ago ) without any errors and the actual performance looked very good. I also run the whole list of tests and the errors it reported were exactly the same as the x86 version. It was something unrelated to coroutines, some UTF8? I might be wrong.

 I'll try to revisit this later this week or in the weekend and make a PR. 

Thank for you work and language btw, it has been one of my "guilty pleasures" since the late 00's  😄

**Comment by stevedekorte** [comment]
Resolved on the `wasm` branch (the new direction for the project).

The root cause was platform-specific register save/restore in `asm.S` (`NEEDARMCONTEXT` / `NEEDPOWERCONTEXT` ifdefs) used by the C-stack-switching coroutine implementation. The stackless rewrite replaces that entirely with heap-allocated `IoEvalFrame` objects and an iterative evaluator:

- `asm.S` is gone
- No `ucontext` / `setjmp` / `longjmp` / `makecontext` anywhere in the VM
- `IoCoroutine.h`: *"There is NO C stack switching (no setjmp/longjmp, no ucontext, no fibers)"*

Coroutines, continuations, and Futures all run on the frame-based eval loop, so the whole class of "won't build on arch X because asm.S doesn't cover it" is eliminated. Apple Silicon, ARM32, and WASM all fall out for free.

Closing as resolved on `wasm`.

**Comment by stevedekorte** [comment]
Correction to my previous comment — the resolution comes from the **stackless** VM rewrite (PR #488), which has already been merged to `master`, not from the wasm work specifically. The asm.S / platform-specific coroutine code is gone on `master` itself. Apple Silicon, ARM32, and any other target that was blocked by missing asm.S ifdefs all build without that obstacle now.
